### PR TITLE
udp: add mmsg support to darwin

### PR DESCRIPTION
**Pull request by res0nance**
adds mmsg support to darwin platform using internal sendmsg_x / recvmsg_x functions

### CODE REVIEW

**Comment by saghul** [comment]
Ha, nice one! The header file says they are private APIs not to be relied upon, but it looks like they haven't changed in years!

**Review by saghul — APPROVED** [review]

File: src/unix/darwin-syscalls.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef UV_DARWIN_SYSCALLS_H_
+#define UV_DARWIN_SYSCALLS_H_
+
+#include <sys/types.h>
+#include <sys/socket.h>
+
+// https://github.com/apple/darwin-xnu/blob/master/bsd/sys/socket.h
```
Comment (bnoordhuis) [inline src/unix/darwin-syscalls.h:7]:
```suggestion
/* https://github.com/apple/darwin-xnu/blob/master/bsd/sys/socket.h */
```

File: src/unix/darwin-syscalls.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+struct mmsghdr {
+    struct msghdr msg_hdr;
+    size_t msg_len;
+};
+
+ssize_t recvmsg_x(int s, const struct mmsghdr *msgp, u_int cnt, int flags);
+ssize_t sendmsg_x(int s, const struct mmsghdr *msgp, u_int cnt, int flags);
```
Comment (bnoordhuis) [inline src/unix/darwin-syscalls.h:15]:
```suggestion
ssize_t recvmsg_x(int s, const struct mmsghdr* msgp, u_int cnt, int flags);
ssize_t sendmsg_x(int s, const struct mmsghdr* msgp, u_int cnt, int flags);
```

File: src/unix/udp.c [hunk 3]
```diff
@@ -173,11 +173,18 @@ static int uv__udp_recvmmsg(uv_udp_t* handle, uv_buf_t* buf) {
     msgs[k].msg_hdr.msg_control = NULL;
     msgs[k].msg_hdr.msg_controllen = 0;
     msgs[k].msg_hdr.msg_flags = 0;
+    msgs[k].msg_len = 0;
```
Comment (bnoordhuis) [inline src/unix/udp.c:176]: For my curiosity, doesn't xnu like it when the field is uninitialized?

**Review by bnoordhuis — APPROVED** [review]

**Comment by barracuda156** [comment]
I realize that perhaps nobody cares, but this has broken the build on some macOS versions.
```
Undefined symbols:
  "_sendmsg_x", referenced from:
      _uv__udp_sendmsg_many in libuv.a(udp.o)
  "_recvmsg_x", referenced from:
      _uv__udp_io in libuv.a(udp.o)
ld: symbol(s) not found
```
And impossibility to fix this here forces to carry patches for every port which bundles its own `libuv` copy for some reason (for instance, some R packages do)... :(

**Comment by res0nance** [comment]
> I realize that perhaps nobody cares, but this has broken the build on some macOS versions.
> 
> ```
> Undefined symbols:
>   "_sendmsg_x", referenced from:
>       _uv__udp_sendmsg_many in libuv.a(udp.o)
>   "_recvmsg_x", referenced from:
>       _uv__udp_io in libuv.a(udp.o)
> ld: symbol(s) not found
> ```
> 
> And impossibility to fix this here forces to carry patches for every port which bundles its own `libuv` copy for some reason (for instance, some R packages do)... :(

What versions of MacOS does this break?

**Comment by i3roly** [comment]
everything beneath 10.15 iirc.

but's okay.

ARM RULZ. flavour of the day. let's all get stupid and use bad processors!

**Comment by res0nance** [comment]
> everything beneath 10.15 iirc.
> 
> but's okay.
> 
> ARM RULZ. flavour of the day. let's all get stupid and use bad processors!

Thanks, since 10.15 and below are not supported platforms I won't consider this an issue https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md

**Pull request merged**